### PR TITLE
Subscription Rework and multiple version support

### LINKED ISSUE #24: Extract workspace subscriptions from catalog methods and move to extended methods.
Catalog methods should only contain 1 to 1 mappings to the catalog framework API.

### PR DESCRIPTION
**Pull request by rzwiefel**
#### Remaining to do/possible changes
- [x] Check if these changes are 2.24/2.25 compatible and change any of the `2.26` nomenclature to 2.24+/2.25+ depending on which is supported.
- [x] Update root readme with information about email-service and how to install the different versions (feature repo add, feature install)
- [x] A complete runthrough of all registered methods on both a 2.19 and 2.26 instance to ensure no runtime issues were missed
- [x] miscellaneous code cleanup (see comments i left on pr)

------
Closes #24 

### Overview
- Subscriptions are no longer coupled with the ddf.catalog/query response (Did not match what the catalog framework does)
  - Two new Subscriptions methods: One for checking if you are subscribed to the supplied ids, and one for fetching all your current subscriptions
  - ⚠️ This will break existing users of query method relying on subscriptions (working with downstream consumers to fix)
- Added support for multiple versions of DDF: 2.19, 2.26+
  - Abstracted version incompatibilities away behind a service and support into two implementations of that service that can be installed with the corresponding versions via feature install. 
- New feature module with Feature file containing the two previously mentioned features: `jsonrpc-2.19` and `jsonrpc-2.26`

### Bundle changes
-  ddf-rpc-api 
  - Custom Import package utilizing version ranges that will work with both 2.19 and 2.26 (mostly ddf packages, but also shiro and geotools

#### New Bundles:
- email-service 
  - Contains the EmailResolver interface and exports it
- 2.19-email-impl & 2.26-email-impl
  - Contains implementations for EmailResolver for their respective versions

### Blueprint changes
- ddf-rpc-api
  - No longer uses SubjectOperations(Service)/SecurityUtils(Static class methods) -- Now uses `email-service`
- 2.19-email-impl/2.26-email-impl
  - Registers the EmailResovler implementation as a service

### Structural changes
- No longer a single module, so we now have a few levels of modules
image

### CODE REVIEW

File: ddf-rpc-api/pom.xml [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+                        <Export-Package>
+                            com.connexta.jsonrpc
+                        </Export-Package>
+                        <Embed-Dependency>
+                            catalog-core-api-impl,
+                            platform-util,
+                            log-sanitizer
+                        </Embed-Dependency>
+                        <!--
+                        org.codice.ddf.log.sanitizer;version="[2.19,3)",
```
Comment (rzwiefel) [inline ddf-rpc-api/pom.xml:145]: remove

File: ddf-rpc-api/pom.xml [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+                            !com.google.errorprone.annotations.*,
+                            !javax.annotation,
+                            !org.checkerframework.checker.nullness.qual,
+                            org.opengis.annotation;version="[19.1.0,21.0.0)",
+                            org.opengis.filter;version="[19.1.0,21.0.0)",
+                            org.opengis.filter.sort;version="[19.1.0,21.0.0)",
+                            org.opengis.filter.expression;version="[19.1.0,21.0.0)",
+                            org.apache.commons.lang3;version="[3.8.0,4.0.0)",
+                            org.apache.commons.lang3.tuple;version="[3.8.0,4.0.0)",
+                            ddf.security;version="[2.19,3)",
```
Comment (rzwiefel) [inline ddf-rpc-api/pom.xml:158]: See if this is removeable or if the transitive dependency powers that be require us specify this

File: ddf-rpc-api/pom.xml [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+                            ddf.catalog.transform;version="[2.19,3)",
+                            ddf.catalog.util;version="[2.19,3)",
+                            ddf.catalog.validation;version="[2.19,3)",
+                            ddf.catalog.validation.report;version="[2.19,3)",
+                            ddf.catalog.validation.violation;version="[2.19,3)",
+                            org.apache.shiro;version="[1.4.0,2.0.0)",
+                            org.apache.shiro.subject;version="[1.4,2)",
+                            *
+                        </Import-Package>
+                        <DynamicImport-Package>ddf.security, ddf.security.impl</DynamicImport-Package>
```
Comment (rzwiefel) [inline ddf-rpc-api/pom.xml:186]: Confirm this is no longer needed

File: ddf-rpc-api/src/main/java/com/connexta/ddf/persistence/subscriptions/SubscriptionMethods.java [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    }
+
+    Set<String> subscriptions =
+        new HashSet<>(getSubscriptions(emailResolver.getCurrentSubjectEmail()));
+
+    Map<String, Boolean> resultMap =
+        ids.stream()
+            .collect(
+                Collectors.toMap(Function.identity(), subscriptions::contains, (l, r) -> l || r));
+    return resultMap; // TODO (RCZ) - in progress
```
Comment (rzwiefel) [inline ddf-rpc-api/src/main/java/com/connexta/ddf/persistence/subscriptions/SubscriptionMethods.java:99]: remove todo

File: ddf-rpc-api/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 5]
```diff
@@ -60,7 +63,7 @@
     <argument ref="listHandler" />
   </bean>
 
-  <reference id="subjectOperations" interface="ddf.security.SubjectOperations" />
+<!--  <reference id="subjectOperations" interface="ddf.security.SubjectOperations" />-->
```
Comment (rzwiefel) [inline ddf-rpc-api/src/main/resources/OSGI-INF/blueprint/blueprint.xml:66]: remove

File: ddf-rpc-api/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 6]
```diff
@@ -71,7 +74,7 @@
     <argument ref="subscriptionMethods"/>
     <argument ref="metacardMap" />
     <argument ref="listHandler" />
-    <argument ref="subjectOperations" />
+<!--    <argument ref="subjectOperations" />-->
```
Comment (rzwiefel) [inline ddf-rpc-api/src/main/resources/OSGI-INF/blueprint/blueprint.xml:77]: remove

File: email/2.19-email-impl/pom.xml [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+                <groupId>org.apache.felix</groupId>
+                <artifactId>maven-bundle-plugin</artifactId>
+                <version>3.5.1</version>
+                <extensions>true</extensions>
+                <configuration>
+                    <instructions>
+                        <Bundle-SymbolicName>${project.artifactId}</Bundle-SymbolicName>
+                        <Export-Package />
+
+<!--                        <Import-Package>-->
```
Comment (rzwiefel) [inline email/2.19-email-impl/pom.xml:51]: remove

File: email/2.19-email-impl/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 8]
```diff
@@ -0,0 +1,21 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+    /**
+    * Copyright (c) Codice Foundation
```
Comment (rzwiefel) [inline email/2.19-email-impl/src/main/resources/OSGI-INF/blueprint/blueprint.xml:4]: codice -> cx

File: email/2.26-email-impl/pom.xml [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+                <groupId>org.apache.felix</groupId>
+                <artifactId>maven-bundle-plugin</artifactId>
+                <version>3.5.1</version>
+                <extensions>true</extensions>
+                <configuration>
+                    <instructions>
+                        <Bundle-SymbolicName>${project.artifactId}</Bundle-SymbolicName>
+                        <Export-Package />
+
+<!--                        <Import-Package>-->
```
Comment (rzwiefel) [inline email/2.26-email-impl/pom.xml:51]: remove

**Comment by rzwiefel** [comment]
Tested all methods against 2.19 and master (currently 2.27-SNAPSHOT) successfully

File: README.md [hunk 10]
```diff
@@ -13,3 +13,13 @@ network. It should have the following properties:
   to enable more efficient communication when possible.
 - **Documented.** The [list-methods](https://localhost:8993/direct) rpc
   method should list out all known methods.
+
+## Installing on different versions
+
+Currently there is support for 2.18 and 2.25+ (this may run on versions between, but has not been tested).
```
Comment (gjvera) [inline README.md:19]: ❓ Should this be 2.19 instead of 2.18?

Comment (rzwiefel) [reply README.md:19]: yes this should be 2.19, good catch. I suspect it will work for quite a few versions past 2.19, but i haven't tested them. Primarily was making sure master + 2.18 were good.

File: ddf-rpc-api/src/main/java/com/connexta/ddf/persistence/subscriptions/SubscriptionMethods.java [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+    METHODS = builder;
+  }
+
+  public SubscriptionMethods(PersistentStore persistentStore, EmailResolver emailResolver) {
+    this.persistentStore = persistentStore;
+    this.emailResolver = emailResolver;
+  }
+
+  public Object getAllSubscriptions(Map<String, Object> params) {
+    return MapFactory.mapOf(
```
Comment (Bdthomson) [inline ddf-rpc-api/src/main/java/com/connexta/ddf/persistence/subscriptions/SubscriptionMethods.java:72]: What does `MapFactory` provide?

**Review by Bdthomson — APPROVED** [review]

File: ddf-rpc-api/src/main/java/com/connexta/ddf/persistence/subscriptions/SubscriptionMethods.java (hunk 11 above)
Comment (rzwiefel) [reply ddf-rpc-api/src/main/java/com/connexta/ddf/persistence/subscriptions/SubscriptionMethods.java:72]: Its what we use instead of using guava for terse construction of maps.  I pulled out guava in the previous PR if you're interested.

**Comment by rzwiefel** [comment]
Verified that this works with ddf 2.23.2 as well

File: README.md [hunk 12]
```diff
@@ -13,3 +13,13 @@ network. It should have the following properties:
...
+
+## Installing on different versions
+
+Currently there is support for 2.19 and 2.25+ (this may run on versions between, but has not been tested).
+In order to install a specific version into a running distribution, build or install these artifacts locally, then run:
+
+`feature:repo-add mvn:com.connexta.jsonrpc/jsonrpc-features/0.5-SNAPSHOT/xml/features`
+
+
+This will add the two features for the previously mentioned features. Run `feature:install jsonrpc-2.19` or `feature:install jsonrpc-2.25` depending on which version you need.
```
Comment (jrnorth) [inline README.md:25]: ✏️ I don't understand "the two features for the previously mentioned features"

File: ddf-rpc-api/pom.xml [hunk 13]
```diff
@@ -0,0 +1,292 @@
...
+        <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-deploy-plugin</artifactId>
+                <version>${maven.deploy.plugin.version}</version>
+            </plugin>
+            <plugin>
+                <groupId>org.apache.felix</groupId>
+                <artifactId>maven-bundle-plugin</artifactId>
+                <version>3.5.1</version>
```
Comment (jrnorth) [inline ddf-rpc-api/pom.xml:131]: This is a pretty old version. DDF is on 4.2.1

File: ddf-rpc-api/pom.xml [hunk 14]
```diff
@@ -0,0 +1,292 @@
...
+                            org.opengis.filter;version="[19.1.0,21.0.0)",
+                            *
+                        </Import-Package>
+                    </instructions>
+                </configuration>
+            </plugin>
+            <plugin>
+                <groupId>com.coveo</groupId>
+                <artifactId>fmt-maven-plugin</artifactId>
+                <version>2.3.0</version>
```
Comment (jrnorth) [inline ddf-rpc-api/pom.xml:191]: ✏️ DDF is on 2.9.1, but if you want to leave a reformat to another PR then that's ok

File: ddf-rpc-api/pom.xml [hunk 15]
```diff
@@ -0,0 +1,292 @@
...
+                <configuration>
+                    <autoVersionSubmodules>true</autoVersionSubmodules>
+                    <preparationGoals>clean verify install</preparationGoals>
+                    <pushChanges>false</pushChanges>
+                </configuration>
+            </plugin>
+            <plugin>
+                <groupId>org.jacoco</groupId>
+                <artifactId>jacoco-maven-plugin</artifactId>
+                <version>0.8.2</version>
```
Comment (jrnorth) [inline ddf-rpc-api/pom.xml:214]: ✏️ 0.8.6 is the latest

File: ddf-rpc-api/src/main/java/com/connexta/ddf/catalog/direct/CatalogMethods.java [hunk 16]
```diff
@@ -485,25 +478,22 @@ private boolean isWorkspace(Metacard metacard) {
...
 
-  private List<Object> getResults(
-      QueryResponse queryResponse, List<String> workspaceSubscriptionIds) {
+  private List<Object> getResults(QueryResponse queryResponse) {
     return queryResponse
         .getResults()
         .stream()
         .map(Result::getMetacard)
-        .map(metacard -> getMetacardInfo(metacard, workspaceSubscriptionIds))
+        .map(metacard -> getMetacardInfo(metacard))
```
Comment (jrnorth) [inline ddf-rpc-api/src/main/java/com/connexta/ddf/catalog/direct/CatalogMethods.java:496]: ✏️ Can use a method reference instead

File: ddf-rpc-api/src/main/java/com/connexta/ddf/persistence/subscriptions/SubscriptionMethods.java [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+
+  private static final String ID = "id";
+
+  public static final String IDS = "ids";
+
+  private final PersistentStore persistentStore;
+
+  private final EmailResolver emailResolver;
+
+  private final Map<String, RpcMethod> METHODS;
```
Comment (jrnorth) [inline ddf-rpc-api/src/main/java/com/connexta/ddf/persistence/subscriptions/SubscriptionMethods.java:49]: ✏️ Should be `methods` since it's not `static final`

File: ddf-rpc-api/src/main/java/com/connexta/ddf/persistence/subscriptions/SubscriptionMethods.java [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+
+  private final RpcFactory rpc = new RpcFactoryImpl();
```
Comment (jrnorth) [inline ddf-rpc-api/src/main/java/com/connexta/ddf/persistence/subscriptions/SubscriptionMethods.java:51]: ✏️ Consider injecting this instead

File: ddf-rpc-api/src/main/java/com/connexta/ddf/persistence/subscriptions/SubscriptionMethods.java [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+
+  {
+    Map<String, RpcMethod> builder = new HashMap<>();
+    builder.put(
```
Comment (jrnorth) [inline ddf-rpc-api/src/main/java/com/connexta/ddf/persistence/subscriptions/SubscriptionMethods.java:55]: ✏️ Could just add directly to `methods` in the constructor

File: ddf-rpc-api/src/main/java/com/connexta/ddf/persistence/subscriptions/SubscriptionMethods.java [hunk 20]
```diff
@@ -0,0 +1,145 @@
...
+  }
+
+  public Object getAllSubscriptions(Map<String, Object> params) {
+    return MapFactory.mapOf(
+        "subscriptions", getSubscriptions(emailResolver.getCurrentSubjectEmail()));
+  }
+
+  public Object getSubscriptions(Map<String, Object> params) {
+
+    if (!params.containsKey(IDS) || !(params.get(IDS) instanceof List)) {
```
Comment (jrnorth) [inline ddf-rpc-api/src/main/java/com/connexta/ddf/persistence/subscriptions/SubscriptionMethods.java:78]: ✏️ `if (!(params.get(IDS) instanceof List))` is sufficient

File: ddf-rpc-api/src/main/java/com/connexta/ddf/persistence/subscriptions/SubscriptionMethods.java [hunk 21]
```diff
@@ -0,0 +1,145 @@
...
+            .collect(
+                Collectors.toMap(Function.identity(), subscriptions::contains, (l, r) -> l || r));
+    return MapFactory.mapOf("isSubscribed", resultMap);
+  }
+
+  private List<Map<String, Object>> query(String email) throws PersistenceException {
+    String queryableEmail = "\"" + EMAIL_PROPERTY + "\"=" + "'" + email + "'";
+    LOCK.lock();
+    try {
+      List<Map<String, Object>> results =
```
Comment (jrnorth) [inline ddf-rpc-api/src/main/java/com/connexta/ddf/persistence/subscriptions/SubscriptionMethods.java:107]: ✏️ Can return directly

File: email/2.19-email-impl/pom.xml [hunk 22]
```diff
@@ -0,0 +1,142 @@
...
+        <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-deploy-plugin</artifactId>
+                <version>${maven.deploy.plugin.version}</version>
+            </plugin>
+            <plugin>
+                <groupId>org.apache.felix</groupId>
+                <artifactId>maven-bundle-plugin</artifactId>
+                <version>3.5.1</version>
```
Comment (jrnorth) [inline email/2.19-email-impl/pom.xml:44]: ✏️ Same comments here w.r.t. plugin versions

File: ddf-rpc-api/pom.xml [hunk 23]
```diff
@@ -0,0 +1,292 @@
...
+                        <goals>
+                            <goal>prepare-agent</goal>
+                        </goals>
+                    </execution>
+                    <execution>
+                        <id>default-report</id>
+                        <phase>prepare-package</phase>
+                        <configuration>
+                            <outputDirectory>
+                                ${project.build.directory}/site/${project.report.output.directory}/jacoco/
```
Comment (jrnorth) [inline ddf-rpc-api/pom.xml:234]: ✏️ I think `${project.report.output.directory}` is a custom property so make sure it's defined somewhere

File: email/2.19-email-impl/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 24]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+    /**
+    * Copyright (c) Connexta
+    **/
+-->
+<blueprint xmlns="http://www.osgi.org/xmlns/blueprint/v1.0.0" xmlns:cm="http://aries.apache.org/blueprint/xmlns/blueprint-cm/v1.1.0">
+
+  <bean id="emailImpl" class="com.connexta.jsonrpc.email.impl.v_2_19.EmailImpl">
+  </bean>
```
Comment (jrnorth) [inline email/2.19-email-impl/src/main/resources/OSGI-INF/blueprint/blueprint.xml:10]: ✏️ Self-closing tag

File: email/2.25-email-impl/pom.xml [hunk 25]
```diff
@@ -0,0 +1,142 @@
...
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-release-plugin</artifactId>
+                <version>${maven.release.plugin.version}</version>
+                <configuration>
+                    <autoVersionSubmodules>true</autoVersionSubmodules>
+                    <preparationGoals>clean verify install</preparationGoals>
+                    <pushChanges>false</pushChanges>
+                </configuration>
+            </plugin>
+            <plugin>
```
Comment (jrnorth) [inline email/2.25-email-impl/pom.xml:76]: ✏️ Actually, we may want to put some of these plugins in the parent POM so we don't have to copy so much

File: jsonrpc-features/pom.xml [hunk 26]
```diff
@@ -0,0 +1,85 @@
...
+        <groupId>com.connexta</groupId>
+        <artifactId>ddf-jsonrpc</artifactId>
+        <version>0.5-SNAPSHOT</version>
+    </parent>
+
+    <groupId>com.connexta.jsonrpc</groupId>
+    <artifactId>jsonrpc-features</artifactId>
+    <packaging>pom</packaging>
+    <name>Connexta :: JSON RPC :: Features</name>
+		<dependencies>
```
Comment (jrnorth) [inline jsonrpc-features/pom.xml:17]: ✏️ Formatting

File: README.md (hunk 12 above)
Comment (rzwiefel) [reply README.md:25]: Indeed, It doesn’t make a whole lot of sense. Lets see, I think i’ll change it to something along the lines of:
> The `feature:repo-add` will add the features for jsonrpc available in your system. Run `Feature:install`......

Does that sound a little better?

File: ddf-rpc-api/src/main/java/com/connexta/ddf/persistence/subscriptions/SubscriptionMethods.java (hunk 18 above)
Comment (rzwiefel) [reply ddf-rpc-api/src/main/java/com/connexta/ddf/persistence/subscriptions/SubscriptionMethods.java:51]: Hmm, so I probably put this here just out of laziness. Makes sense to pass it in though, particularly to help testing. Will change 👌

File: email/2.25-email-impl/pom.xml (hunk 25 above)
Comment (rzwiefel) [reply email/2.25-email-impl/pom.xml:76]: 92AF70F3-6A7D-4499-83C3-DE3DC91405ED
I was avoiding this todo list item 😂 . Will do though. Probably will put these changes in the next PR (that’s already open)

File: ddf-rpc-api/pom.xml (hunk 13 above)
Comment (rzwiefel) [reply ddf-rpc-api/pom.xml:131]: Updated. And moved to plugin management so it is centralized. (in #33 )

File: ddf-rpc-api/pom.xml (hunk 14 above)
Comment (rzwiefel) [reply ddf-rpc-api/pom.xml:191]: Yup i'll update and include it in the next PR (#33)

File: ddf-rpc-api/pom.xml (hunk 15 above)
Comment (rzwiefel) [reply ddf-rpc-api/pom.xml:214]: Updated and moved to plugin management (in #33)

File: ddf-rpc-api/pom.xml (hunk 23 above)
Comment (rzwiefel) [reply ddf-rpc-api/pom.xml:234]: I removed for now since i'm not sure what it was doing for us anyways.

File: README.md (hunk 12 above)
Comment (jrnorth) [reply README.md:25]: 👍

**Comment by rzwiefel** [comment]
CI build passed ✅ 🎉
